src/hooks/resources/device.ts

```diff
@@ -494,3 +509,64 @@ addDeleteHookForDependents('device', [
 	['service_install', 'device'],
 	['gateway_download', 'is_downloaded_by__device'],
 ]);
+
+async function checkSupervisorReleaseUpgrades(
+	api: PinejsClient,
+	deviceIds: number[],
+	newSupervisorReleaseId: number,
+) {
+	const newSupervisorRelease = (await api.get({
+		resource: 'supervisor_release',
+		id: newSupervisorReleaseId,
+		options: {
+			$select: 'supervisor_version',
+		},
+	})) as AnyObject;
+
+	if (newSupervisorRelease == null) {
+		throw new BadRequestError(
+			`Could not find a supervisor release with this ID ${newSupervisorReleaseId}`,
+		);
+	}
+
+	const newSupervisorVersion = newSupervisorRelease.supervisor_version;
+
+	const devices = (await api.get({
+		resource: 'device',
+		options: {
+			$filter: {
+				id: {
+					$in: deviceIds,
+				},
+			},
+			$expand: {
+				should_be_managed_by__supervisor_release: {
+					$select: ['supervisor_version'],
+				},
+			},
+		},
+	})) as AnyObject[];
```

Comment:
It would be better to invert this as
```suggestion
	const devices = (await api.get({
		resource: 'supervisor_release',
		options: {
			$select: ['supervisor_version'],
			$filter: {
				should_manage__device: {
					$any: {
						$alias: 'd',
						$expr: {
							d: {
								id: {
									$in: deviceIds,
								},
							}.
						},
					},
				},
			},
		},
	})) as AnyObject[];
```
and avoid fetching a load of wasteful data (it'll also simplify a bunch of the checks below)